### ISSUE TITLE
Ban Try/Catch

### ISSUE THREAD

**Issue by osyrisrblx** [roblox-ts/roblox-ts#873]
There have been several "try catch reworks" in the past:
#101 #171 #335 

Initially we tried to allow object throw + catch, but we realized there wasn't a safe way to do this.

What we've finally landed on is compiling try/catch as:
```TS
try {
    print("foo");
} catch (e) {
    print(e);
}
```

```Lua
local _0, _1 = pcall(function()
	print("foo");
end);
if not _0 then
	local e = _1;
	print(e);
end;
```

**However, this is bad for a number of reasons:**
1. We have to ban usage of `return`, because that would happen within `pcall(function() ... end)` and not actually return the outer function.
2. We have an outstanding bug where you can use `break` inside of try/catch but that causes invalid Lua with the error `no loop to break near ';'`.
3. We have an outstanding bug where you can use `continue` inside of try/catch and causes the same issues as `break`.
4. Our implementation of `finally` is incorrect! The `finally` block should run even if the `try` or `catch` blocks return the outer function or throw. However, even if we could fix the above issues, `finally` is basically impossible to model in Lua.

Alternatively, you could write the above code as just:
```TS
const result = opcall(() => print("foo"));
if (!result.success) {
    print(result.error);
}
```

Which I think is much less misleading. It's clear why you can't `return`, `break`, or `continue` inside of that function call. It also lets you model return values much better via `result.value`.

### So with all of the above stated, I propose that we ban try/catch from being used in roblox-ts. I think it causes more harm than it does good.

However, we would not ban `throw "some string";` because that can easily just compile to `error("some string");` and it useful for representing dead code paths.

**Comment by osyrisrblx** [comment]
I'd like to add that _maybe_ we could work around some of the problems I stated. However, even if try/catch _was_ perfect, it's still just an abstraction over pcall for syntax sugar. I think using opcall is much more direct and easier to reason about.

**Comment by plt-hokusai** [comment]
I have a fix for this, though I'm not familiar with your compiler specifically, I'm assuming that you have a way to change how statements are compiled depending on the semantic context they appear in (Here, I'm assuming that you can change how `break;`, `continue;`, and `return;` are compiled within the scope of the `try { }` block).

You can define compilation for JS-style
```js
try {
  /* ... k_try ... */
} catch(e) {
  /* ... k_catch ... */
} finally {
  /* ... k_finally ... */
}
```
blocks in a way that preserves semantics at the cost of being pretty.

First, introduce program-scoped sentinels `_BREAK`, `_CONTINUE` and `_RETURN`. All of the variables I'm prefixing with a `_` are taken to be locally fresh, but local freshness in the global scope is the same thing as global freshness, anyway; We're going to use these to compile `break;`, `continue;` and `return ...`; statements in the scope of a `try` block.

Compiling `try`
===========

1. Compile `break` to `coroutine.yield(_BREAK)`.
2. Compile `continue` to `coroutine.yield(_CONTINUE)`.
3. Compile `return x, ..., z` to `return _RETURN, x, ..., z`.

Now that you have the compiled statements for inside the `try`-block, we're going to shove them in a coroutine, as became clear with the use of `yield` to implement _BREAK and _CONTINUE. Coroutines give us a clean way to section off execution of a program fragment and decide what to do before, during, and after its execution.

Create a coroutine for the code that executes within a `try` block:

```lua
local _c = coroutine.create(function()
  --- k_try
end)
```

For example, the code fragment
```js
try {
  let x = /* read some number */
  if (x == 2) {
    break;
  } else {
    throw "x wasn't two"
  }
} catch(e) { /* more stuff here */ }
```
gives rise to the code fragment
```lua
local _c = coroutine.create(function()
  local x = --[[ read some number ]]--
  if x == 2 then
    coroutine.yield(_BREAK)
  else
    error("x wasn't two")
  end
end
-- more stuff
```

Compiling `catch`
=============

Now that we have our coroutine `_c` we can run it for a step (or many; more on this later) and see how it reacts.

To compile a `catch(e) { /* handler stuff */ }` block (after compiling its respective `try`, of course), generate this code:

```lua
local _r = coroutine.resume(c)
if not _r[1] then
  -- handler stuff
elseif _r[1] and _r[2] == _BREAK then
  break
elseif _r[1] and _r[2] == _CONTINUE then
  continue
elseif _r[1] and _r[2] == _RETURN then
  return table.unpack(_r, 3, #_r)
elseif #_r >= 2 then
  coroutine.yield(table.unpack(_r, 2, #_r))
end
```

Compiling `finally`
==============

Compile:
```js
/* try stuff */ } finally {
  print("a")
  return 1;
}
```
into
```lua
local function _finally()
  print("a")
  return 1
end
```

Then paste the `finally` exit handlers at the appropriate point before the `catch` code exits the scope:

```lua
local _r = coroutine.resume(c)
if not _r[1] then
  do
    -- handler stuff
  end
  -- finally code here
elseif _r[1] then
  -- finally code here
  if _r[2] == _BREAK then
    break
  elseif _r[2] == _CONTINUE then
    continue -- however you lot do this
  elseif _r[2] == _RETURN then
    return table.unpack(_r, 3, #_r)
  elseif #_r >= 1 then
    coroutine.yield(table.unpack(_r, 2, #_r))
  end
end
```

I've bunched up the common exit paths together so that the continuation doesn't need to be quadruplicated, unfortunately it does still need to be duplicated. There is a tradeoff here between duplicating the size of the `finally` handler and implementing it with the same coroutine kludge used to handle non-local exits from the `try` block.

Why?
====

This addresses all of the points you raised in the issue description.

> 1. We have to ban usage of return, because that would happen within pcall(function() ... end) and not actually return the outer function.

Now you don't; pcall is never involved, and by making use of a sentinel value, you can tell whether the function yielded because of a return statement, because of a call to yield, or by exiting normally.

> 2. We have an outstanding bug where you can use break inside of try/catch but that causes invalid Lua with the error no loop to break near ';'.
> 3. We have an outstanding bug where you can use continue inside of try/catch and causes the same issues as break.

Again, using sentinel values tells the code handling the `try` block to continue or break.

> 4. Our implementation of finally is incorrect! The finally block should run even if the try or catch blocks return the outer function or throw. However, even if we could fix the above issues, finally is basically impossible to model in Lua.

:tada:

**Comment by Dionysusnu** [comment]
@plt-hokusai Something similar to this has now been implemented on the refactor branch, in #1020 .
There is no need to wrap in a coroutine, a simple xpcall is enough. This is handled with the RuntimeLib's `TS.try(func, catch, finally)`
